### ISSUE TITLE
Crash if listen ports are already ocupied

### ISSUE THREAD

**Issue by ikoveshnikov** [tempesta-tech/tempesta#721]
Steps to reproduce:

1. Start anything, that bind 0.0.0.0:80
2. Start Tempesta with same listen settings.
3. Have a crash.

It is a user responsibility to start Tempesta on a free ports, but having a kernel crash is too heavy price for the mistake.

```
[  613.141023] [tdb] Start Tempesta DB
[  613.149095] [tempesta] Initializing Tempesta FW kernel module...
[  613.155431] [tempesta] Registering new scheduler: hash
[  613.158642] [tempesta] Registering new scheduler: http
[  613.162892] [tempesta] Registering new scheduler: round-robin
[  613.168535] [tempesta] Starting all modules...
[  613.177408] [tdb] Opened table /opt/tempesta/db/filter.tdb: size=16777216 rec_size=20 base=ffff9527fc400000
[  613.178770] [tempesta] Open listen socket on: 0.0.0.0
[  613.179615] [tempesta] ERROR: can't bind to: 0.0.0.0
[  613.180989] [tempesta] ERROR: can't start listening on: 0.0.0.0
[  613.182703] [tempesta] ERROR: start() for module 'sock_clnt' returned the error: -98
[  613.198630] [tdb] Close table 'filter0.tdb'
[  613.199589] ------------[ cut here ]------------
[  613.200011] kernel BUG at /home/user/tempesta/tempesta/tempesta_fw/connection.h:433!
[  613.200011] invalid opcode: 0000 [#1] SMP
[  613.200011] Modules linked in: tfw_sched_rr(OE) tfw_sched_http(OE) tfw_sched_hash(OE) tempesta_fw(OE) tempesta_db(OE) tempesta_tls(OE) crct10dif_pclmul(E) crc32_pclmul(E) iTCO_wdt(E) iTCO_vendor_support(E) ghash_clmulni_intel(E) ppdev(E) snd_hda_codec_generic(E) evdev(E) pcspkr(E) serio_raw(E) qxl(E) snd_hda_intel(E) snd_hda_codec(E) snd_hda_core(E) snd_hwdep(E) snd_pcm(E) snd_timer(E) snd(E) lpc_ich(E) soundcore(E) mfd_core(E) sg(E) ttm(E) drm_kms_helper(E) virtio_balloon(E) virtio_console(E) drm(E) shpchp(E) parport_pc(E) parport(E) acpi_cpufreq(E) tpm_tis(E) tpm_tis_core(E) tpm(E) button(E) ip_tables(E) x_tables(E) autofs4(E) ext4(E) crc16(E) jbd2(E) crc32c_generic(E) fscrypto(E) ecb(E) mbcache(E) sr_mod(E) cdrom(E) crc32c_intel(E) virtio_blk(E) aesni_intel(E) aes_x86_64(E) glue_helper(E) lrw(E) gf128mul(E) ablk_helper(E) cryptd(E) ehci_pci(E) ahci(E) uhci_hcd(E) libahci(E) ehci_hcd(E) e1000(E) i2c_i801(E) i2c_smbus(E) psmouse(E) libata(E) sym53c8xx(E) scsi_transport_spi(E) usbcore(E) usb_common(E) scsi_mod(E) virtio_pci(E) virtio_ring(E) virtio(E)
[  613.200011] CPU: 0 PID: 6121 Comm: sysctl Tainted: G           OE   4.8.0-tempesta-amd64 #1 Debian 4.8.15-tfw1-1
[  613.200011] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.10.2-20170228_101828-anatol 04/01/2014
[  613.200011] task: ffff9527d9a9e100 task.stack: ffff9527d998c000
[  613.200011] RIP: 0010:[<ffffffffc068a138>]  [<ffffffffc068a138>] tfw_sock_srv_del_conns+0x108/0x120 [tempesta_fw]
[  613.200011] RSP: 0018:ffff9527d998fd78  EFLAGS: 00010203
[  613.200011] RAX: 0000000000000001 RBX: ffff9527d9211dc8 RCX: dead000000000200
[  613.200011] RDX: ffff9527dc123840 RSI: ffff9527d9956400 RDI: ffff9527dc123850
[  613.200011] RBP: ffff9527d9211d90 R08: ffff9527d9211dc8 R09: 0000000000000005
[  613.200011] R10: 0000000000100001 R11: ffffffffbe2a446d R12: ffff9527dc123850
[  613.200011] R13: ffff9527dc123840 R14: ffff9527d92116c8 R15: ffff9527d998ff20
[  613.200011] FS:  00007f35f985c880(0000) GS:ffff9527ffc00000(0000) knlGS:0000000000000000
[  613.200011] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  613.200011] CR2: 00007f35f8f57d20 CR3: 0000000059520000 CR4: 00000000003406f0
[  613.200011] Stack:
[  613.200011]  ffff9527dc123840 ffff9527d9956090 ffffffffc068a030 ffff9527d9956080
[  613.200011]  0000563dc56532e0 ffffffffc0686bad ffff9527d998ff20 ffffffffc069eac0
[  613.200011]  ffffffffc069ea80 00000000ffffff9e ffffb6b840462000 ffffffffc0689de5
[  613.200011] Call Trace:
[  613.200011]  [<ffffffffc068a030>] ? tfw_sock_srv_start+0x1e0/0x1e0 [tempesta_fw]
[  613.200011]  [<ffffffffc0686bad>] ? tfw_sg_for_each_srv+0x5d/0xa0 [tempesta_fw]
[  613.200011]  [<ffffffffc0689de5>] ? tfw_clean_srv_groups+0x15/0x60 [tempesta_fw]
[  613.200011]  [<ffffffffc066bfae>] ? tfw_cfg_start+0x8e/0x1c0 [tempesta_fw]
[  613.200011]  [<ffffffffc0685756>] ? handle_sysctl_state_io+0xf6/0x1d0 [tempesta_fw]
[  613.200011]  [<ffffffffbd3f12ce>] ? mutex_lock+0xe/0x30
[  613.200011]  [<ffffffffc068569f>] ? handle_sysctl_state_io+0x3f/0x1d0 [tempesta_fw]
[  613.200011]  [<ffffffffbd0789d3>] ? proc_sys_call_handler+0xe3/0x100
[  613.200011]  [<ffffffffbd0011e0>] ? vfs_write+0xb0/0x190
[  613.200011]  [<ffffffffbd0025b2>] ? SyS_write+0x52/0xc0
[  613.200011]  [<ffffffffbd3f38f6>] ? system_call_fast_compare_end+0xc/0x96
[  613.200011] Code: 8b a6 a0 00 00 00 4d 85 e4 74 13 49 8b 56 38 4c 89 f5 49 89 c6 48 39 d3 0f 85 45 ff ff ff 0f 0b 5b 31 c0 5d 41 5c 41 5d 41 5e c3 <0f> 0b 0f 0b 0f 0b 0f 0b 0f 0b 0f 0b 0f 0b 66 2e 0f 1f 84 00 00 
[  613.200011] RIP  [<ffffffffc068a138>] tfw_sock_srv_del_conns+0x108/0x120 [tempesta_fw]
[  613.200011]  RSP <ffff9527d998fd78>
[  613.279703] ---[ end trace f27777ffdc8a6882 ]---
[  613.286000] [tempesta] Un-registering scheduler: hash
[  613.412266] kmem_cache_destroy tfw_srv_conn_cache: Slab cache still has objects
[  613.414406] CPU: 0 PID: 6134 Comm: rmmod Tainted: G      D    OE   4.8.0-tempesta-amd64 #1 Debian 4.8.15-tfw1-1
[  613.416008] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.10.2-20170228_101828-anatol 04/01/2014
[  613.416008]  0000000000000286 0000000063b43ab0 ffffffffbd127054 ffff9527d958a3c0
[  613.416008]  0000000000000000 ffffffffbcfa691d 00ff9527d958a5a0 ffff9527d9993e70
[  613.416008]  ffff9527d9993e70 ffff9527d9993e80 ffff9527d9993e80 0000000063b43ab0
[  613.416008] Call Trace:
[  613.416008]  [<ffffffffbd127054>] ? dump_stack+0x5c/0x78
[  613.416008]  [<ffffffffbcfa691d>] ? kmem_cache_destroy+0x1fd/0x210
[  613.416008]  [<ffffffffc0685860>] ? tfw_exit+0x30/0x60 [tempesta_fw]
[  613.416008]  [<ffffffffbcefee8c>] ? SyS_delete_module+0x18c/0x260
[  613.416008]  [<ffffffffbce03290>] ? exit_to_usermode_loop+0xa0/0xc0
[  613.416008]  [<ffffffffbd3f38f6>] ? system_call_fast_compare_end+0xc/0x96
[  618.317430] BUG: unable to handle kernel NULL pointer dereference at 0000000000000010
[  618.321364] IP: [<ffffffffc066c8a7>] tfw_connection_drop+0x17/0x30 [tempesta_fw]
[  618.321364] PGD 0 
[  618.321364] Oops: 0000 [#2] SMP
[  618.321364] Modules linked in: tempesta_fw(OE-) tempesta_db(OE) tempesta_tls(OE) crct10dif_pclmul(E) crc32_pclmul(E) iTCO_wdt(E) iTCO_vendor_support(E) ghash_clmulni_intel(E) ppdev(E) snd_hda_codec_generic(E) evdev(E) pcspkr(E) serio_raw(E) qxl(E) snd_hda_intel(E) snd_hda_codec(E) snd_hda_core(E) snd_hwdep(E) snd_pcm(E) snd_timer(E) snd(E) lpc_ich(E) soundcore(E) mfd_core(E) sg(E) ttm(E) drm_kms_helper(E) virtio_balloon(E) virtio_console(E) drm(E) shpchp(E) parport_pc(E) parport(E) acpi_cpufreq(E) tpm_tis(E) tpm_tis_core(E) tpm(E) button(E) ip_tables(E) x_tables(E) autofs4(E) ext4(E) crc16(E) jbd2(E) crc32c_generic(E) fscrypto(E) ecb(E) mbcache(E) sr_mod(E) cdrom(E) crc32c_intel(E) virtio_blk(E) aesni_intel(E) aes_x86_64(E) glue_helper(E) lrw(E) gf128mul(E) ablk_helper(E) cryptd(E) ehci_pci(E) ahci(E) uhci_hcd(E) libahci(E) ehci_hcd(E) e1000(E) i2c_i801(E) i2c_smbus(E) psmouse(E) libata(E) sym53c8xx(E) scsi_transport_spi(E) usbcore(E) usb_common(E) scsi_mod(E) virtio_pci(E) virtio_ring(E) virtio(E) [last unloaded: tfw_sched_rr]
[  618.321364] CPU: 1 PID: 0 Comm: swapper/1 Tainted: G      D    OE   4.8.0-tempesta-amd64 #1 Debian 4.8.15-tfw1-1
[  618.321364] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.10.2-20170228_101828-anatol 04/01/2014
[  618.321364] task: ffff9527dc8ea000 task.stack: ffff9527dc90c000
[  618.321364] RIP: 0010:[<ffffffffc066c8a7>]  [<ffffffffc066c8a7>] tfw_connection_drop+0x17/0x30 [tempesta_fw]
[  618.321364] RSP: 0018:ffff9527ffd03b80  EFLAGS: 00010246
[  618.321364] RAX: 0000000000000000 RBX: ffff9527d9211ba0 RCX: 0000000000000030
[  618.321364] RDX: 0000000000000004 RSI: ffff9527ffd03b9a RDI: ffff9527d9211ba0
[  618.321364] RBP: ffff9527d99e9800 R08: 0000000000000320 R09: 0000000000000000
[  618.321364] R10: 0000000000000000 R11: 0000000000000002 R12: ffff9527dc20dc00
[  618.321364] R13: ffff9527da381c62 R14: 0000000000000001 R15: ffff9527d99e9888
[  618.321364] FS:  0000000000000000(0000) GS:ffff9527ffd00000(0000) knlGS:0000000000000000
[  618.321364] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  618.321364] CR2: 0000000000000010 CR3: 000000005beaa000 CR4: 00000000003406e0
[  618.321364] Stack:
[  618.321364]  ffff9527d9211ba0 ffffffffc068a253 2e30312e302e3031 0000303030383a31
[  618.321364]  0000000000000000 0000000000000000 0000000000000000 0000000000000000
[  618.321364]  ffff000000000000 5f65491f09bfe022 ffff9527d99e9800 ffff9527d99e9800
[  618.321364] Call Trace:
[  618.321364]  <IRQ> 
[  618.321364]  [<ffffffffc068a253>] ? tfw_sock_srv_connect_failover+0x73/0xf0 [tempesta_fw]
[  618.321364]  [<ffffffffc0687e0b>] ? ss_linkerror+0x2b/0x60 [tempesta_fw]
[  618.321364]  [<ffffffffbd3481ae>] ? tcp_validate_incoming+0x27e/0x370
[  618.321364]  [<ffffffffbd3492eb>] ? tcp_rcv_established+0x19b/0x6c0
[  618.321364]  [<ffffffffbd3070ad>] ? sk_filter_trim_cap+0x2d/0x2c0
[  618.321364]  [<ffffffffbd353a83>] ? tcp_v4_do_rcv+0x133/0x1f0
[  618.321364]  [<ffffffffbd3553ec>] ? tcp_v4_rcv+0x8ac/0xa60
[  618.321364]  [<ffffffffbceb433e>] ? find_busiest_group+0x3e/0x4d0
[  618.321364]  [<ffffffffbd32eed9>] ? ip_local_deliver_finish+0x99/0x1c0
[  618.321364]  [<ffffffffbd32f19b>] ? ip_local_deliver+0x6b/0xf0
[  618.321364]  [<ffffffffbd354b30>] ? tcp_v4_early_demux+0x130/0x140
[  618.321364]  [<ffffffffbd32ebc1>] ? ip_rcv_finish+0x171/0x3f0
[  618.321364]  [<ffffffffbd32f4a6>] ? ip_rcv+0x286/0x3e0
[  618.321364]  [<ffffffffbceb499c>] ? load_balance+0x1cc/0xa00
[  618.321364]  [<ffffffffbd2ef109>] ? __netif_receive_skb_core+0x4f9/0xa10
[  618.321364]  [<ffffffffbd2f07e3>] ? process_backlog+0x93/0x130
[  618.321364]  [<ffffffffbd2efed0>] ? net_rx_action+0x240/0x370
[  618.321364]  [<ffffffffbd3f6423>] ? __do_softirq+0x113/0x2a6
[  618.321364]  [<ffffffffbce7f7eb>] ? irq_exit+0x9b/0xa0
[  618.321364]  [<ffffffffbd3f5ae2>] ? call_function_single_interrupt+0x82/0x90
[  618.321364]  <EOI> 
[  618.321364]  [<ffffffffbce59012>] ? native_safe_halt+0x2/0x10
[  618.321364]  [<ffffffffbce2f6da>] ? default_idle+0x1a/0xd0
[  618.321364]  [<ffffffffbcebb7a8>] ? cpu_startup_entry+0x2e8/0x340
[  618.321364]  [<ffffffffbce47efd>] ? start_secondary+0x14d/0x190
[  618.321364] Code: 85 c0 74 06 48 83 c4 08 ff e0 48 83 c4 08 c3 0f 1f 44 00 00 0f 1f 44 00 00 8b 47 10 53 48 89 fb 83 e0 0f 48 8b 04 c5 c0 44 8c c0 <48> 8b 40 10 48 85 c0 74 02 ff d0 48 83 bb 98 00 00 00 00 75 02 
[  618.321364] RIP  [<ffffffffc066c8a7>] tfw_connection_drop+0x17/0x30 [tempesta_fw]
[  618.321364]  RSP <ffff9527ffd03b80>
[  618.321364] CR2: 0000000000000010
[  618.321364] ---[ end trace f27777ffdc8a6883 ]---
```

**Comment by keshonok** [comment]
The bug is reproducible and confirmed. To reproduce the bug, the configuration must have at least one back end server configured.

**Comment by keshonok** [comment]
Fixed via a541e3d.
